Adding a capability to store chat histories to Azure Cosmos DB via NoSQL API

rioriost opened the pull request:
A draft implementation to store chat history to Azure Cosmos DB instead of a JSON file. If you have a time for review, I'm happy to fix my codes to meet your code policies.

rioriost committed:
Create cosmos.py

To store chat history to Azure Cosmos for NoSQL

rioriost committed:
Update __init__.py

Import CosmosConversationMemory from cosmos.py

rioriost committed:
Update README.md

Add a sample to store chat history to Cosmos DB for NoSQL

taikinman commented on src/langrila/memory/cosmos.py:
```diff
@@ -0,0 +1,45 @@
+from ..base import BaseConversationMemory
+
+class CosmosConversationMemory(BaseConversationMemory):
+    def __init__(self, endpoint_name: str, key_name: str, dbname_name: str, containername_name: str, pkey_name: str):
```
Do you have any concern if you change containername_name to containername?

taikinman commented on src/langrila/__init__.py:
```diff
@@ -9,6 +9,7 @@
 )
 from .embedding.openai import OpenAIEmbeddingModule
 from .memory.json import JSONConversationMemory
+from .memory.cosmos import CosmosConversationMemory
```
It could be better not to import in __init__.py because additional dependencies are needed.

rioriost replied:
Got it. It should be moved to each Python scripts to use Cosmos.

taikinman replied:
Yes, but it's OK.
I forgot to suggest.

```suggestion
```

taikinman commented:
@rioriost 

Sorry for the late and thank you for the first contribution!!
I added two comments above, so could you check it?

taikinman commented on src/langrila/memory/cosmos.py:
```diff
@@ -0,0 +1,45 @@
+        self.key = os.getenv(key_name)
+        self.dbname = os.getenv(dbname_name)
+        self.containername = os.getenv(containername_name)
+        self.pkey = '/' + os.getenv(pkey_name).strip('/')
```
What is pkey used for? and what's the input? (sorry I'm not familiar with cosmosdb)

rioriost replied:
Partition key is used to distribute the data into physical partitions of Cosmos DB backend.
I'm acceptable to change 'pkey' to 'partition_key' to clarify the meaning.

taikinman replied:
I see, but partition_key seems not to be used in this module right now and not always required. Thus it might be better if we can choose whether we use partition_key or not. I wonder we can use partition_key as the argument of `self.container.create_item()`.

rioriost replied:
Sorry, you're right. I forgot to remove 'pkey' during fixing the logic in 'store()'.

taikinman commented on src/langrila/memory/cosmos.py:
```diff
@@ -0,0 +1,45 @@
+    def load(self) -> list[dict[str, str]]:
+        result = []
+        try:
+            items = self.container.read_items()
```
```suggestion
    items = self.container.read_all_items()
```

rioriost replied:
It's enough reasonable to change from read_items() to read_all_items().

taikinman commented on src/langrila/memory/cosmos.py:
```diff
@@ -0,0 +1,45 @@
+from ..base import BaseConversationMemory
+
+class CosmosConversationMemory(BaseConversationMemory):
+    def __init__(self, endpoint_name: str, key_name: str, dbname_name: str, containername_name: str, pkey_name: str):
```
```suggestion
def __init__(self, endpoint_name: str, key_name: str, dbname_name: str, container_name: str):
```

rioriost replied:
It's a little hard to understand that the name of these arguments mean the name of env variables. But I can't judge what is the best way to fix. Because I'd like to align the naming rule that you use in other modules.
Are you acceptable a rule as follows?
endpoint_name -> endpoint_env_name, key_name -> key_anv_name,...

taikinman replied:
> It's a little hard to understand that the name of these arguments mean the name of env variables. 

You're exactly right. It is problem of my rough coding design.

> endpoint_name -> endpoint_env_name, key_name -> key_anv_name,...

Sure, I feel it's better rule for readability. Thanks! Also I will change argument names of other env variables used at other modules like from `api_key_name` to `api_key_env_name`.

taikinman commented on src/langrila/memory/cosmos.py:
```diff
@@ -0,0 +1,45 @@
+        try:
+            items = self.container.read_items()
+            for item in items:
+                result.append(item)
```
Later step needs only `role` and `content`. 

```suggestion
result.append({k: v for k, v in item.items() if k in {"role", "content"}})
```

rioriost replied:
Cosmos DB returns not only 'role' and 'content' actually. e.g. 'id', '_rid', '_etag', '_ts', and so on.
Especially, '_ts', timestamp is important to sort the conversation history.

But I can also understand your intention to keep your langrila small and clean. Any ideas?

taikinman replied:
Sorry, I don't quite understand your intention precisely yet. I'd appreciate it if you could give me more details.
How can we use those metadatas from CosmosDB? Or how will those metadatas be used at other place out of this module.

Basically I'm trying to close the responsibility of each module into each module itself. If we need to sort conversation memories with `_ts `and if that's all to do, it might be enough to use `_ts` for sorting memories within `load()` and to output the list of message dictionary including only `role` and `content`.
But I also think I couldn't understand due to my poor knowledge on CosmosDB. Thus I need more details.

rioriost replied:
I understand your intention to close the responsibility. load() in CosmosConversationMemory should be implemented exactly same to one in JSONConversationMemory.
If a user needs to use other values from Cosmos DB, he/she can inherit CosmosConversationMemory class and add load_all() or something to get all the values from Cosmos. It's clean.

taikinman commented on src/langrila/memory/cosmos.py:
```diff
@@ -0,0 +1,45 @@
+
+    def store(self, conversation_history: list[dict[str, str]]):
+        for item in conversation_history:
+            item['id'] = str(uuid.uuid4())
+            self.container.create_item(item)
```
To avoid duplication of items.

```suggestion
            item["id"] = hashlib.sha256(str(item).encode()).hexdigest()
            try:
                self.container.create_item(item)
            except exceptions.CosmosResourceExistsError:
                pass
```

rioriost replied:
nice solution!
Thanks!

taikinman commented:
@rioriost 

sorry for the repost.
Thank you for the first contribution!!
I have reviewed, so could you check it?

rioriost commented:
Great Thanks, taikinman-san!

I TRULY agreed with your blog entry not to use LangChain.
If I can contribute langrila a little, it's pleasure for me!

taikinman commented on README.md:
````diff
@@ -234,6 +234,28 @@ response.model_dump()
+## Conversation memory with Cosmos DB
+
+```python
+from langrila import CosmosConversationMemory
````
following the change of \_\_init\_\_.py

```suggestion
from langrila.memory.cosmos import CosmosConversationMemory
```

taikinman commented on src/langrila/memory/cosmos.py:
```diff
@@ -0,0 +1,45 @@
+from azure.cosmos import exceptions, CosmosClient, PartitionKey
+import logging
+import os
+import uuid
+
```
```suggestion
import hashlib
```

taikinman commented:
> Great Thanks, taikinman-san!
> 
> I TRULY agreed with your blog entry not to use LangChain.
> If I can contribute langrila a little, it's pleasure for me!

Wow, thank you so much for visiting my blog entry!
It motivates me and I'm pleased to be able to make langrila better with you!

rioriost committed:
Update README.md

Moved import from __init__.py to each Python scripts to avoid adding dependencies to langrila itself.

rioriost committed:
Update __init__.py

Removed import to avoid adding dependencies.

rioriost committed:
Update cosmos.py

Changed the name of arguments to clarify their role to point environment variables
Changed uuid to hashlib
Changed return values to align to JSONConversationMemory class

rioriost committed:
Update README.md

Changed arguments name to clarify the role to point environment variables

rioriost committed:
Update README.md

taikinman reviewed (approved):
LGTM.
Nice contribution!
